src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
 
             if (customAttributeCtorParameters.Length != 0 || customAttributeNamedParameters.Length != 0)
             {
-                ParseAttributeArguments(
-                    attributeBlob.Signature,
-                    (int)attributeBlob.Length,
-                    ref customAttributeCtorParameters,
-                    ref customAttributeNamedParameters,
-                    (RuntimeAssembly)customAttributeModule.Assembly);
+                CustomAttributeDataParser parser = new CustomAttributeDataParser(attributeBlob);
+                try
+                {
+                    if (!parser.ValidateProlog())
+                    {
+                        throw new BadImageFormatException(SR.Arg_CustomAttributeFormatException);
+                    }
+
+                    ParseCtorArgs(ref parser, ref customAttributeCtorParameters, customAttributeModule);
+                    ParseNamedArgs(ref parser, ref customAttributeNamedParameters, customAttributeModule);
+                }
+                catch (Exception ex) when (ex is not OutOfMemoryException)
+                {
+                    throw new CustomAttributeFormatException(ex.Message, ex);
+                }
             }
         }
 
-        public CustomAttributeType CustomAttributeType => m_type;
-        public long PrimitiveValue => m_primitiveValue;
-        public CustomAttributeEncodedArgument[] ArrayValue => m_arrayValue;
-        public string StringValue => m_stringValue;
-    }
+        internal CustomAttributeEncodedArgument(CustomAttributeType type)
+        {
+            CustomAttributeType = type;
+        }
 
-    [StructLayout(LayoutKind.Auto)]
-    internal readonly struct CustomAttributeNamedParameter
-    {
-        private readonly string m_argumentName;
-        private readonly CustomAttributeEncoding m_fieldOrProperty;
-        private readonly CustomAttributeEncoding m_padding;
-        private readonly CustomAttributeType m_type;
-        private readonly CustomAttributeEncodedArgument m_encodedArgument;
+        public CustomAttributeType CustomAttributeType { get; private set; }
+        public PrimitiveValue PrimitiveValue { get; set; }
+        public CustomAttributeEncodedArgument[]? ArrayValue { get; set; }
+        public string? StringValue { get; set; }
 
-        public CustomAttributeNamedParameter(string argumentName, CustomAttributeEncoding fieldOrProperty, CustomAttributeType type)
+        private static void ParseCtorArgs(
+            ref CustomAttributeDataParser parser,
+            ref CustomAttributeCtorParameter[] customAttributeCtorParameters,
```

Comment:
```suggestion
            CustomAttributeCtorParameter[] customAttributeCtorParameters,
```
Does not need to be `ref`?